File: Behavior/Mediator/Go/mediator.go
package main

import "fmt"

// Mediator interface
type Mediator interface {
	Send(message string, colleague Colleague)
}

// Colleague interface
type Colleague interface {
	Send(message string)
	Receive(message string)
	SetMediator(mediator Mediator)
}

// ConcreteMediator is a mediator for Colleague objects
type ConcreteMediator struct {
	colleague1 Colleague
	colleague2 Colleague
}

// Send message to all colleagues except the sender
func (m *ConcreteMediator) Send(message string, sender Colleague) {
	if m.colleague1 == sender {
		m.colleague2.Receive(message)
	} else {
		m.colleague1.Receive(message)
	}
}

// SetColleague1 sets the first colleague
func (m *ConcreteMediator) SetColleague1(colleague Colleague) {
	m.colleague1 = colleague
}

// SetColleague2 sets the second colleague
func (m *ConcreteMediator) SetColleague2(colleague Colleague) {
	m.colleague2 = colleague
}

// ConcreteColleague1 is a concrete colleague
type ConcreteColleague1 struct {
	mediator Mediator
}

// Send message through mediator
func (c *ConcreteColleague1) Send(message string) {
	c.mediator.Send(message, c)
}

// Receive message from mediator
func (c *ConcreteColleague1) Receive(message string) {
	fmt.Println("Colleague1 Received:", message)
}

// SetMediator sets the mediator
func (c *ConcreteColleague1) SetMediator(mediator Mediator) {
	c.mediator = mediator
}

// ConcreteColleague2 is a concrete colleague
type ConcreteColleague2 struct {
	mediator Mediator
}

// Send message through mediator
func (c *ConcreteColleague2) Send(message string) {
	c.mediator.Send(message, c)
}

// Receive message from mediator
func (c *ConcreteColleague2) Receive(message string) {
	fmt.Println("Colleague2 Received:", message)
}

// SetMediator sets the mediator
func (c *ConcreteColleague2) SetMediator(mediator Mediator) {
	c.mediator = mediator
}

// main function
func main() {
	mediator := &ConcreteMediator{}

	colleague1 := &ConcreteColleague1{}
	colleague1.SetMediator(mediator)
	colleague2 := &ConcreteColleague2{}
	colleague2.SetMediator(mediator)

	mediator.SetColleague1(colleague1)
	mediator.SetColleague2(colleague2)

	colleague1.Send("Hi there!")
	colleague2.Send("Hello!")
}

File: Behavior/Mediator/Lua/mediator.lua
-- 中介者类
local Mediator = {}
function Mediator:new()
    local obj = { colleagues = {} }
    self.__index = self
    return setmetatable(obj, self)
end

function Mediator:register(colleague)
    table.insert(self.colleagues, colleague)
end

function Mediator:send(message, sender)
    for _, colleague in ipairs(self.colleagues) do
        if colleague ~= sender then
            colleague:receive(message)
        end
    end
end

-- 同事基类
local Colleague = {}
function Colleague:new(mediator)
    local obj = { mediator = mediator }
    self.__index = self
    return setmetatable(obj, self)
end

function Colleague:send(message)
    self.mediator:send(message, self)
end

function Colleague:receive(message)
    -- 由子类实现
end

-- 具体的同事类
local ConcreteColleague1 = Colleague:new()

function ConcreteColleague1:receive(message)
    print("ConcreteColleague1 received: " .. message)
end

local ConcreteColleague2 = Colleague:new()

function ConcreteColleague2:receive(message)
    print("ConcreteColleague2 received: " .. message)
end

-- 使用中介者模式
local mediator = Mediator:new()

local colleague1 = ConcreteColleague1:new(mediator)
local colleague2 = ConcreteColleague2:new(mediator)

mediator:register(colleague1)
mediator:register(colleague2)

colleague1:send("Hello")
colleague2:send("Good day")
